include/pybind11/functional.h

```diff
@@ -75,6 +84,36 @@ struct type_caster<std::function<Return(Args...)>> {
             // PYPY segfaults here when passing builtin function like sum.
             // Raising an fail exception here works to prevent the segfault, but only on gcc.
             // See PR #1413 for full details
+        } else {
```

Comment:
To try to minimize the runtime overhead, I would consider adding a rec->next check here to only trigger this if there is a another possible overload to try.

Reply:
I'm trying to understand your comment but I'm afraid I don't know what you are suggesting. Can you please elaborate?